### PR TITLE
`tarfile.open()`: Handle all modes

### PR DESCRIPTION
**Pull request by srittau**

### CODE REVIEW

**Review by srittau — COMMENTED** [review]
Add pipes

**Comment by AlexWaygood** [comment]
Looks like you're running into https://github.com/python/mypy/issues/16367. It's been fixed on mypy `master`, but the fix hasn't yet been included in a mypy release.

**Comment by AlexWaygood** [comment]
> Looks like you're running into [python/mypy#16367](https://github.com/python/mypy/issues/16367). It's been fixed on mypy `master`, but the fix hasn't yet been included in a mypy release.

Oh sorry, you already noted as such in https://github.com/python/typeshed/issues/12182#issue-2366140417 :-)

**Comment by srittau** [comment]
> > Looks like you're running into [python/mypy#16367](https://github.com/python/mypy/issues/16367). It's been fixed on mypy `master`, but the fix hasn't yet been included in a mypy release.
> 
> Oh sorry, you already noted as such in [#12182 (comment)](https://github.com/python/typeshed/issues/12182#issue-2366140417) :-)

I sneak edited it in when I noticed the problem. (Which had already been noted by @max-muoto in #12117.)

**Comment by srittau** [comment]
vision is doing some dynamic construction of the mode string. This should be easily solvable by using literals and an intermediate variable. Which sounds like a good idea to ensure that no invalid mode strings are constructed.

pip is assigning the mode string using an if/elif cascade. This could also be solved by using an intermediate variable.

Both are not great, but I'm personally fine with these, given the increases soundness of the annotations. (Of course, in pip's case it would be even better if mypy would automatically infer that `mode` is a literal with certain values.)

**Comment by JelleZijlstra** [comment]
This has a merge conflict.

File: stdlib/tarfile.pyi [hunk 1]
```diff
@@ -121,16 +119,124 @@ def open(
     pax_headers: Mapping[str, str] | None = ...,
     debug: int | None = ...,
     errorlevel: int | None = ...,
-    compresslevel: int | None = ...,
```
Comment (JelleZijlstra) [inline stdlib/tarfile.pyi:124]: I see this argument being used in the implementation, why did you remove it?

Comment (srittau) [reply stdlib/tarfile.pyi:124]: Due to this comment from the docs:

> For modes 'w:gz', 'x:gz', 'w|gz', 'w:bz2', 'x:bz2', 'w|bz2', [tarfile.open()](https://docs.python.org/3/library/tarfile.html#tarfile.open) accepts the keyword argument compresslevel (default 9) to specify the compression level of the file.
